pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -898,14 +899,35 @@ func possibleGuestSize(disk api.Disk) (int64, bool) {
 		return 0, false
 	}
 
-	size := int64((1 - filesystemOverhead) * float64(preferredSize))
+	usableSize, err := getUsableDiskSize(getSourceFile(disk))
+	if err != nil {
+		log.DefaultLogger().Reason(err).Error("Failed to get total usable space, using disk capacity instead")
+		usableSize = *disk.Capacity
+	}
+
+	size := int64((1 - filesystemOverhead) * float64(usableSize))
 	size = kutil.AlignImageSizeTo1MiB(size, log.DefaultLogger())
 	if size == 0 {
 		return 0, false
 	}
 	return size, true
 }
 
+func getUsableDiskSize(path string) (int64, error) {
+	var stat syscall.Statfs_t
+	err := syscall.Statfs(path, &stat)
+	if err != nil {
+		return int64(-1), err
+	}
+	availableSize := int64(stat.Bavail) * int64(stat.Bsize)
+
+	diskInfo, err := converter.GetImageInfo(path)
```

Comment:
I think you can propagate `diskInfo` from `shouldExpandOffline`

Reply:
`possibleGuestSize` is used in several places that don't use diskInfo so I think it makes more sense to have it here.

Reply:
I worry a little about a second call to `qemu-img info`.. maybe just use a stat syscall?

Reply:
wdyt about this?

Reply:
Sorry I forgot about this, why should we be concerned with `qemu-img` calls? I'll try syscall though

Reply:
I haven't found any way to get the actual size `qemu-img` provides with stat or statfs.

Reply:
should be those two multiplied, or, maybe the info.Size() https://pkg.go.dev/os#File.Stat
```
$ qemu-img info tests/images/cirros-qcow2.img --output json | grep actual
                "actual-size": 12374016,
    "actual-size": 12374016,
$ stat -c %B tests/images/cirros-qcow2.img
512
$ stat -c %b tests/images/cirros-qcow2.img
24168
```
https://man7.org/linux/man-pages/man1/stat.1.html

Reply:
> I haven't found any way to get the actual size `qemu-img` provides with stat or statfs.

In fact after checking again, stat seems to provide a `size` field that matches quemu-img's actualSize... using that instead. Sorry for the confussion.

Reply:
In that case, you can (probably) even use the golang file API, something like `fileinfo, err := f.Stat(); fileinfo.Size()`